esp-hal/src/spi/master.rs

```diff
@@ -2113,6 +2113,11 @@ where
 
         rx.is_done();
 
+        // disable the write side when reading is being done
+        reg_block
+            .user()
+            .modify(|_, w| w.usr_miso().bit(true).usr_mosi().bit(false));
+
```

Comment:
The HIL tests won't catch this but now that MOSI is being disabled for reads, it also has to be re-enabled when a write or transfer is done afterwards. As it is now, writes will be disabled until the user re-creates the driver.

Reply:
Ideally we should add a test for this case

Reply:
I'm trying to reproduce it and here is the interesting part of the code I'm testing (the init part is copied from spi_loopback_dma example):
```rust
loop {
        send[0] = i;
        send[send.len() - 1] = i;
        i = i.wrapping_add(1);

        let transfer = spi.dma_read(&mut receive).unwrap();

        transfer.wait().unwrap();

        println!(
            "{:x?} .. {:x?}",
            &receive[..10],
            &receive[receive.len() - 10..]
        );

        let transfer = spi.dma_write(&mut send).unwrap();

        transfer.wait().unwrap();

        println!("{:x?} .. {:x?}", &send[..10], &send[send.len() - 10..]);

        delay.delay_millis(250);
    }
```
    
and the result
    
```rust
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0] .. [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9] .. [73, 74, 75, 76, 77, 78, 79, 7a, 7b, 0]
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0] .. [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[1, 1, 2, 3, 4, 5, 6, 7, 8, 9] .. [73, 74, 75, 76, 77, 78, 79, 7a, 7b, 1]
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0] .. [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[2, 1, 2, 3, 4, 5, 6, 7, 8, 9] .. [73, 74, 75, 76, 77, 78, 79, 7a, 7b, 2]
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0] .. [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[3, 1, 2, 3, 4, 5, 6, 7, 8, 9] .. [73, 74, 75, 76, 77, 78, 79, 7a, 7b, 3]
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0] .. [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[4, 1, 2, 3, 4, 5, 6, 7, 8, 9] .. [73, 74, 75, 76, 77, 78, 79, 7a, 7b, 4]
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0] .. [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

it's tested with the current code in this PR (no MOSI re-enablement). I will re-enable it since is is a right thing to do but the result is a bit weird.

Reply:
The code only prints out the slice that was given to the driver to send, you'd need a slave or logical analyser to see to the problem. Oh no actually using `dma_transfer` instead of `dma_write` should show the problem, since that would read what is being written by the hardware.

Reply:
Thank you @Dominaezzz, hope it's resolved now.